README.md
# Modbusd a simple modbus tcp server
Implements read_holding_registers (03) and write_multiple_registers (16).
Units are implemented as dynamic libraries.
Linux GPIO is supported.
# Installation
```
git clone https://github.com/tvesalainen/modbusd.git
cd modbusd
make
debuild -us -uc
sudo dpkg -i ../modbusd_0.1_XXXXX.deb
```
After that you have modbus tcp server running with GPIO support at unit 0.
## Command line
```
./modbusd: invalid option -- 'h'
usage:
        -v                      verbose
        -s                      log only to syslog
        -p <port>               port
        -l <unit>:<lib>:<arg>   add plugin
        -h                      help
```

### Verbose
Extra text is written to stderr
### Port
Modbus port is 502. It is possible to change with -p port.
### Plugin
Add plugin with -l.
Example
```
modbusd -l 0:/usr/lib/gpio.so:/dev/gpiochip0

0 is unit. (0 - 255)
/usr/lib/gpio.so is path to dynamic library
/dev/gpiochip0 is library argument
```
There can be several libraries
## Implementing a unit (library)
See plugin.c and gpio.c and Makefile for compiling.
## GPIO
16 bit address lowest 8 bits are used as gpio line offset.
If address high 8 bits are 1 read is polled. In other words read blocks until gpio pin changes
status.

Example: read line 14
```
read 0x0 0xe // normal read

read 0x0 0xe|0x100 // blocking read


```



debian/changelog
modbusd (0.1) unstable; urgency=low

  * Initial release.

 --  Timo Vesalainen <timo.vesalainen@iki.fi>  Mon, 28 Feb 2022 01:54:36 +0000


gpio.c
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <linux/gpio.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <stdint.h>
#include <getopt.h>
#include <stdlib.h>
#include <sys/poll.h>
#include <arpa/inet.h>

#include "plugin.h"

const char *dev;

int read_registers(__u16 address, __u16 quantity, struct modbus_arg *data);
int poll_register(int fd, __u16 address, struct modbus_arg *data);

int init(struct modbusd_ctx *ctx, const char* arg)
{
	int fd;

	fprintf(stderr, "init arg %s\n", arg);
	dev = arg;

	fd = open(dev, O_RDONLY);
	if (fd < 0)
	{
		ERROR("init of %s failed\n", dev);
		return -1;
	}
	close(fd);
	return 0;
}

int read_holding_registers(int fd, __u16 address, __u16 quantity, struct modbus_arg *data)
{
	int hi;
	int lo;

	hi = address>>8;
	lo = address&0xff;
	switch (hi)
	{
		case 0:
			return read_registers(lo, quantity, data);
			break;
		case 1:
			return poll_register(fd, lo, data);
			break;
		default:
			ERROR("%d hi not supported\n", hi);
			return ILLEGAL_DATA_ADDRESS;
			break;
	}
}
int read_registers(__u16 address, __u16 quantity, struct modbus_arg *data)
{
	int fd;
	int rc;
	int ii;
	struct gpiohandle_request req;
	struct gpiohandle_data dat;

	if (quantity > GPIOHANDLES_MAX)
	{
		ERROR("more than %d handles requested\n", GPIOHANDLES_MAX);
		return ILLEGAL_DATA_VALUE;
	}
	fd = open(dev, O_RDONLY);
	if (fd < 0)
	{
		ERROR("open %s failed\n", dev);
		return -1;
	}
	for (ii=0;ii<quantity;ii++)
	{
		req.lineoffsets[ii] = address + ii;
	}
	req.flags = GPIOHANDLE_REQUEST_INPUT ;
	req.lines = quantity;
	rc = ioctl(fd, GPIO_GET_LINEHANDLE_IOCTL, &req);
	close(fd);
	if (rc < 0)
	{
		ERROR("GPIO_GET_LINEHANDLE_IOCTL of %s failed\n", dev);
		return -1;
	}
	rc = ioctl(req.fd, GPIOHANDLE_GET_LINE_VALUES_IOCTL, &dat);
	if (rc < 0)
	{
		ERROR("GPIOHANDLE_GET_LINE_VALUES_IOCTL of %s failed\n", dev);
		return -1;
	}
	close(req.fd);
	for (ii=0;ii<quantity;ii++)
	{
		data->array_short[ii] = htons(dat.values[ii]);
	}
	return OK;
}

int poll_register(int sock, __u16 address, struct modbus_arg *data)
{
	int fd;
	int rc;
	int ii;
	struct gpiohandle_data dat;
	struct gpioevent_request req;
	struct gpioevent_data event;
	struct pollfd pfd[2];

	fd = open(dev, O_RDONLY);
	if (fd < 0)
	{
		ERROR("open %s failed\n", dev);
		return -1;
	}
	req.lineoffset = address;
	strcpy(req.consumer_label, "modbusd");
	req.handleflags = 0;
	req.eventflags = GPIOEVENT_REQUEST_RISING_EDGE | GPIOEVENT_REQUEST_FALLING_EDGE;
	rc = ioctl(fd, GPIO_GET_LINEEVENT_IOCTL, &req);
	close(fd);
	if (rc < 0)
	{
		ERROR("GPIO_GET_LINEEVENT_IOCTL of %s failed\n", dev);
		return -1;
	}
	pfd[0].fd = sock;
	pfd[0].events = POLLHUP;
	pfd[1].fd = req.fd;
	pfd[1].events = POLLIN;
	rc = poll(pfd, 2, -1);
	if (rc < 0)
	{
		ERROR("poll of %s failed\n", dev);
		return -1;
	}
	if (pfd[1].revents != POLLIN)
	{
		return -1;
	}
	rc = read(req.fd, &event, sizeof(event));
	if (rc < 0)
	{
		ERROR("read event of %s failed\n", dev);
		return -1;
	}
	close(req.fd);
	data->array_short[0] = htons(event.id == GPIOEVENT_EVENT_RISING_EDGE ? 1 : 0);
	return OK;
}

int write_multiple_registers(int sock, __u16 address, __u16 quantity, struct modbus_arg *data)
{
	int fd;
	int rc;
	int ii;
	struct gpiohandle_request req;
	struct gpiohandle_data dat;

	if (quantity > GPIOHANDLES_MAX)
	{
		return ILLEGAL_DATA_VALUE;
	}
	fd = open(dev, O_RDONLY);
	if (fd < 0)
	{
		ERROR("open %s failed\n", dev);
		return -1;
	}
	for (ii=0;ii<quantity;ii++)
	{
		req.lineoffsets[ii] = address + ii;
	}
	req.flags = GPIOHANDLE_REQUEST_OUTPUT;
	req.lines = quantity;
	rc = ioctl(fd, GPIO_GET_LINEHANDLE_IOCTL, &req);
	close(fd);
	if (rc < 0)
	{
		ERROR("GPIO_GET_LINEHANDLE_IOCTL of %s failed\n", dev);
		return -1;
	}
	for (ii=0;ii<quantity;ii++)
	{
		dat.values[ii] = ntohs(data->array_short[ii]);
	}
	rc = ioctl(req.fd, GPIOHANDLE_SET_LINE_VALUES_IOCTL, &dat);
	if (rc < 0)
	{
		ERROR("GPIOHANDLE_SET_LINE_VALUES_IOCTL of %s failed\n", dev);
		return -1;
	}
	close(req.fd);
	return OK;
}


modbus.c
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <pthread.h>
#include <sys/uio.h>

#include "plugin.h"
#include "modbus.h"
#include "modbusd.h"

int readn(int s, void *data, size_t len)
{
	int rc;

	while (len > 0)
	{
		rc = read(s, data, len);
		if (rc < 0)
		{
			ERROR("%m: read rc=%d", rc);
			pthread_exit(NULL);
		}
		len -= rc;
		data += rc;
	}
}
int write_response(int s, struct modbus_tcp *tcp_hdr, void *rsp, size_t rsp_size)
{
	struct iovec io[2];
	int rc;

	tcp_hdr->len = htons(rsp_size + 1);
	io[0].iov_base = tcp_hdr;
	io[0].iov_len = sizeof(struct modbus_tcp);
	io[1].iov_base = rsp;
	io[1].iov_len = rsp_size;
	rc = writev(s, io, 2);
	if (rc < 0)
	{
		ERROR("%m: writev rc=%d", rc);
		pthread_exit(NULL);
	}
}

int write_error(int s, struct modbus_tcp *tcp_hdr, int function, int error)
{
	struct modbus_err err;
	struct iovec io[2];
	int rc;

	err.error_code = ERROR_CODE(function);
	err.exception_code = error;
	tcp_hdr->len = htons(3);
	io[0].iov_base = tcp_hdr;
	io[0].iov_len = sizeof(struct modbus_tcp);
	io[1].iov_base = &err;
	io[1].iov_len = sizeof(err);
	rc = writev(s, io, 2);
	if (rc < 0)
	{
		ERROR("%m: writev rc=%d", rc);
		pthread_exit(NULL);
	}
}

void *modbus(void *v)
{
	int s = *(int*)v;
	int rc;
	int true = 1;
	struct modbus_tcp tcp_hdr;
	struct modbus_req req;
	struct modbus_rsp rsp;
	struct modbus_err err;
	int len;
	struct plugin *plugin;
	struct modbus_arg arg;
	unsigned int address;
	unsigned int quantity;

	VERBOSE("new thread %d\n", s);

	if (setsockopt(s, IPPROTO_TCP, TCP_NODELAY, &true, sizeof(true)) < 0)
	{
		ERROR("TCP_NODELAY failed\n");
		pthread_exit(NULL);
	}
	
	for (;;)
	{
		readn(s, &tcp_hdr, sizeof(tcp_hdr));
		len = ntohs(tcp_hdr.len);
		VERBOSE("tx=%d proto=%d len=%d unit=%d\n", 
			ntohs(tcp_hdr.transaction_id),
			ntohs(tcp_hdr.protocol_id),
			len, 
			tcp_hdr.unit_id);
		readn(s, &req, len-1);
		plugin = plugins[tcp_hdr.unit_id]; 
		if (plugin == NULL)
		{
			write_error(s, &tcp_hdr, req.function, GATEWAY_PATH_UNAVAILABLE);
			continue;
		}
		switch (req.function)
		{
			case 3:	/* Read Holding Registers */
				if (plugin->read_holding_registers != NULL)
				{
					address = ntohs(req.address);
					quantity = ntohs(req.quantity);
					if (quantity > 125)
					{
						write_error(s, &tcp_hdr, req.function, ILLEGAL_DATA_VALUE );
						break;
					}
					rc = plugin->read_holding_registers(s, address, quantity, &arg);
					if (rc == 0)
					{
						rsp.function = req.function;
						rsp.bytes = 2*quantity;
						memcpy(&(rsp.data), &arg, 2*quantity);
						write_response(s, &tcp_hdr, &rsp, 2*quantity+2);
					}
					else
					{
						if (rc > 0)
						{
							write_error(s, &tcp_hdr, req.function, rc );
						}
						else
						{
							ERROR("%m: Read Holding Registers\n");
							write_error(s, &tcp_hdr, req.function, SLAVE_DEVICE_FAILURE );
						}
					}
				}
				else
				{
					write_error(s, &tcp_hdr, req.function, ILLEGAL_FUNCTION );
				}
				break;
			case 16:	/* Write Multiple Registers */
				if (plugin->write_multiple_registers != NULL)
				{
					address = ntohs(req.address);
					quantity = ntohs(req.quantity);
					if (quantity > 125)
					{
						write_error(s, &tcp_hdr, req.function, ILLEGAL_DATA_VALUE );
						break;
					}
					rc = plugin->write_multiple_registers(s, address, quantity, &req.data);
					if (rc == 0)
					{
						write_response(s, &tcp_hdr, &req, 5);
					}
					else
					{
						if (rc > 0)
						{
							write_error(s, &tcp_hdr, req.function, rc );
						}
						else
						{
							ERROR("%m: Write Multiple Registers\n");
							write_error(s, &tcp_hdr, req.function, SLAVE_DEVICE_FAILURE );
						}
					}
				}
				else
				{
					write_error(s, &tcp_hdr, req.function, ILLEGAL_FUNCTION );
				}
				break;
			default:
			write_error(s, &tcp_hdr, req.function, ILLEGAL_FUNCTION );
			break;
		}
	}
}


modbus.h
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef MODBUSTCP_H
#define MODBUSTCP_H

#include <linux/types.h>

#include "plugin.h"

#define ERROR_CODE(func) (func|0x80)

struct modbus_tcp
{
	__u16 transaction_id;
	__u16 protocol_id;	/* 0 = MODBUS protocol */
	__u16 len;		/* Number of following bytes */
	__u8  unit_id;
}__attribute__((packed));

struct modbus_req
{
	__u8  function;
	__u16 address;
	__u16 quantity;
	__u8  bytes;
	struct modbus_arg data;
}__attribute__((packed));

struct modbus_rsp
{
	__u8  function;
	__u8  bytes;
	struct modbus_arg data;
}__attribute__((packed));

struct modbus_err
{
	__u8 error_code;
	__u8 exception_code;
}__attribute__((packed));

void *modbus(void *);

#endif /* MODBUSTCP_H */


plugin.c
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdio.h>
#include <linux/types.h>
#include <arpa/inet.h>

#include "plugin.h"

int size;
__u16 *memory;

int init(struct modbusd_ctx *ctx, const char* arg)
{
	fprintf(stderr, "init arg %s\n", arg);
	size = atoi(arg);
	memory = calloc(size, sizeof(__u16));
	if (memory == NULL)
	{
		perror(arg);
		return -1;
	}
	return 0;
}

int read_holding_registers(int fd, __u16 address, __u16 quantity, struct modbus_arg *data)
{
	int ii;

	if (address+2*quantity > size)
	{
		return ILLEGAL_DATA_ADDRESS;
	}
	for (ii=0;ii<quantity;ii++)
	{
		data->array_short[ii] = htons(memory[address+ii]);
	}
	return OK;
}

int write_multiple_registers(int fd, __u16 address, __u16 quantity, struct modbus_arg *data)
{
	int ii;

	if (address+2*quantity > size)
	{
		return ILLEGAL_DATA_ADDRESS;
	}
	for (ii=0;ii<quantity;ii++)
	{
		memory[address+ii] = ntohs(data->array_short[ii]);
	}
	return OK;
}


plugin.h
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _PLUGIN_H
#define _PLUGIN_H

#include <linux/types.h>
#include <syslog.h>

#define OK 0
#define ILLEGAL_FUNCTION 1
#define ILLEGAL_DATA_ADDRESS 2
#define ILLEGAL_DATA_VALUE 3
#define SLAVE_DEVICE_FAILURE 4
#define ACKNOWLEDGE 5
#define SLAVE_DEVICE_BUSY 6
#define MEMORY_PARITY_ERROR 8
#define GATEWAY_PATH_UNAVAILABLE 0xa
#define GATEWAY_TARGET_DEVICE_FAILED_TO_RESPOND 0xb

#define ERROR(...)  syslog(LOG_USER|LOG_ERR, __VA_ARGS__)

struct modbusd_ctx
{
	int verbose;
};

struct modbus_arg
{
	union
	{
		__u8 array_byte[250];
		__u16 array_short[125];	/* in network byte order - use htons and ntohs */
		__u32 array_int[62];	/* in network byte order - use htonl and ntohl */
	};
}__attribute__((packed));

int init(struct modbusd_ctx *ctx, const char* arg);
int read_holding_registers(int fd, __u16 address, __u16 quantity, struct modbus_arg *data);
int write_multiple_registers(int fd, __u16 address, __u16 quantity, struct modbus_arg *data);
 
#endif /* _PLUGIN_H */


mcp342x.c
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <limits.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <fcntl.h>
#include <linux/i2c-dev.h>
#include <i2c/smbus.h>

#include "plugin.h"

/*
	Address
	9-9	ADC Pi
	8-7	PGA Gain
	6-5	Sample Rate
	4-2	Device Address
	1-0	Channel
*/
#define ADC(x) ((x>>9)&0x1)
#define PGA(x) ((x>>7)&0x3)
#define RATE(x) ((x>>5)&0x3)
#define ADDR(x) ((x>>2)&0x7)
#define CHANNEL(x) (x&0x3)
#define ADDRESS(a) (0xd<<3|a)
#define REGISTER(rdy, c, oc, s, g) (rdy<<7|c<<5|oc<<4|s<<2|g)

struct data
{
	union
	{
		struct 
		{
			__s32 pad:4;
			__s32 value:12;
			__s32 ready:1;
			__s32 channel:2;
			__s32 mode:1;
			__s32 rate:2;
			__s32 pga:2;
		} bits_12;
		struct 
		{
			__s32 pad:2;
			__s32 value:14;
			__s32 ready:1;
			__s32 channel:2;
			__s32 mode:1;
			__s32 rate:2;
			__s32 pga:2;
		} bits_14;
		struct 
		{
			__s32 value:16;
			__s32 ready:1;
			__s32 channel:2;
			__s32 mode:1;
			__s32 rate:2;
			__s32 pga:2;
		} bits_16;
		struct 
		{
			__s32 pad:6;
			__s32 value:18;
			__s32 ready:1;
			__s32 channel:2;
			__s32 mode:1;
			__s32 rate:2;
			__s32 pga:2;
		} bits_18;
	};
};
useconds_t SLEEP[] = {41667, 16667, 66667, 266667};
struct modbusd_ctx *ctx;
char dev[PATH_MAX];

#define VERBOSE(...) if (ctx->verbose) fprintf(stderr, __VA_ARGS__)

int init(struct modbusd_ctx *mctx, const char* arg)
{
	int adapter;
	int fd;

	fprintf(stderr, "init arg %s\n", arg);
	ctx = mctx;
	adapter = atoi(arg);
	snprintf(dev, sizeof(dev), "/dev/i2c-%d", adapter);

	fd = open(dev, O_RDONLY);
	if (fd < 0)
	{
		ERROR("init of %s failed\n", dev);
		return -1;
	}
	close(fd);
	return 0;
}

int read_holding_registers(int sock, __u16 address, __u16 quantity, struct modbus_arg *data)
{
	__u8 addr = ADDRESS(ADDR(address));
	__u8 reg = REGISTER(1, CHANNEL(address), 0, RATE(address), PGA(address));
	int fd;
	int rc;
	struct data dat;

	VERBOSE("adc=%d pga=%d rate=%d addr=%d ch=%d\n",
		ADC(address),
		PGA(address),
		RATE(address),
		ADDR(address),
		CHANNEL(address));
	VERBOSE("addr=%x reg=%x\n", addr, reg);

	fd = open(dev, O_RDWR);
	if (fd < 0)
	{
		ERROR("open %s failed\n", dev);
		return -1;
	}
	if  (ioctl(fd, I2C_SLAVE, addr) < 0)
	{
		ERROR("I2C_SLAVE %s failed\n", dev);
		return -1;
	}
	if (i2c_smbus_write_byte(fd, reg) < 0)
	{
		ERROR("i2c_smbus_write_byte_data %s failed\n", dev);
		return -1;
	}
	if (usleep(SLEEP[RATE(address)]) < 0)
	{
		ERROR("usleep %s failed\n", dev);
		return -1;
	}
	rc = read(fd, &dat, sizeof(dat));
	if (rc < 0)
	{
		ERROR("read %s failed\n", dev);
		return -1;
	}

	VERBOSE("%d %x\n", dat.bits_12.value, dat.bits_12.ready);
}


modbusd.h
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef _MODBUSD_H
#define _MODBUSD_H

#include "plugin.h"

#define MODBUS_PORT 502

extern struct modbusd_ctx ctx;

#define VERBOSE(...) if (ctx.verbose) fprintf(stderr, __VA_ARGS__)

struct plugin
{
	int (*read_holding_registers)(int, __u16, __u16, struct modbus_arg*);
	int (*write_multiple_registers)(int, __u16, __u16, struct modbus_arg*);
};

extern struct plugin *plugins[256];

#endif /* _MODBUSD_H */


modbusd.c
/*
 * Copyright (C) 2022 Timo Vesalainen <timo.vesalainen@iki.fi>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <syslog.h>

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <sys/ioctl.h>
#include <arpa/inet.h>
#include <pthread.h>
#include <dlfcn.h>

#include "modbus.h"
#include "modbusd.h"

struct modbusd_ctx ctx;
int use_stderr = 1;

struct plugin *plugins[256];

int add_plugin(const char* arg)
{
	int unit;
	char *parg;
	char *lib;
	char *larg;
	void *plib;
	int (*init)(struct modbusd_ctx*, const char*);

	parg = strdup(arg);
	unit = atoi(parg);
	lib = index(parg, ':');
	if (lib == NULL)
	{
		fprintf(stderr, "illegal unit %s\n", parg);
		exit(-1);
	}
	lib++;
	larg = index(lib, ':');
	if (larg != NULL)
	{
		*larg = 0;
		larg++;
		VERBOSE("add unit %d lib %s %s\n", unit, lib, larg);
	}
	else
	{
		VERBOSE("add unit %d lib %s\n", unit, lib);
	}
	plib = dlopen(lib, RTLD_LAZY);
	if (plib == NULL)
	{
		fprintf(stderr, "couldn't load %s: %s\n", lib, dlerror() );
		exit(-1);
	}

	struct plugin *plugin = (struct plugin*)calloc(1, sizeof(plugin));
	plugins[unit] = plugin;
	plugin->read_holding_registers = dlsym(plib, "read_holding_registers");
	plugin->write_multiple_registers = dlsym(plib, "write_multiple_registers");
	init = dlsym(plib, "init");
	if (init != NULL)
	{
		init(&ctx, larg);
	}
}

int main(int argc, char **argv)
{
	int c;
	int port = MODBUS_PORT;
        int true = 1;
        int false = 0;
	int s;
	int r;
	struct sockaddr_in laddr;
	struct sockaddr_in raddr;
	int size;
	pthread_t thread;

        while ((c = getopt (argc, argv, "vp:sl:")) != -1)
        {
                switch (c)
                {
                        case 'v':
                                ctx.verbose = 1;
                        break;
                        case 's':
                                use_stderr = 0;
                        break;
                        case 'p':
                                port = atoi(optarg);
                        break;
                        case 'l':
                                add_plugin(optarg);
                        break;
                        case 'h':
                        default:
                                fprintf(stderr, "usage:\n"
						"\t-v\t\t\tverbose\n"
						"\t-s\t\t\tlog only to syslog\n"
						"\t-p <port>\t\tport\n"
						"\t-l <unit>:<lib>:<arg>\tadd plugin\n"
						"\t-h\t\t\thelp\n"
				);
                                exit(-1);
                }
        }
        if (use_stderr)
        {
                openlog(NULL, LOG_PERROR, LOG_USER);
        }
        else
        {
                openlog(NULL, 0, LOG_USER);
        }

        VERBOSE("socket\n");
        s = socket(AF_INET, SOCK_STREAM, 0);
        if (s < 0)
        {
                ERROR("%m: socket");
                return -1;
        }
        VERBOSE("setsockopt SO_REUSEADDR\n");
        if(setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &true, sizeof(true)) < 0)
        {
                ERROR("%m: SO_REUSEADDR");
                return -1;
        }
        memset(&laddr, 0, sizeof(laddr));
        laddr.sin_family = AF_INET;
        laddr.sin_addr.s_addr = htonl(INADDR_ANY);
        laddr.sin_port = htons(port);
        VERBOSE("bind\n");
        if (bind(s, (struct sockaddr *)&laddr, sizeof(struct sockaddr)) < 0)
        {
                ERROR("%m: bind");
                return -1;
        }
        VERBOSE("listen\n");
        if (listen(s, 16) < 0)
        {
                ERROR("%m: listen");
                return -1;
	}
	for (;;)
	{
		size = sizeof(struct sockaddr);
		r = accept(s, (struct sockaddr *)&raddr, &size);
        	if (r < 0)
		{
			ERROR("%m: accept");
			return -1;
		}
		VERBOSE("accepted %s:%d\n", inet_ntoa(raddr.sin_addr), ntohs(raddr.sin_port));
		pthread_create(&thread, NULL, modbus, (void*)&r);
	}
}
